### ISSUE TITLE
Open Source Hot Code Push / Over The Air Updates

### ISSUE THREAD

**Issue by alexcroox** [capacitor-community/proposals#43]
## Plugin Request

**Name**: Capacitor Hot Code Push
**Package**: `@capacitor-community/hot-code-push`

### Platform(s)
Android, iOS

### Existing Solutions
MS Code Push - Free, very popular for Cordova, no Capacitor support yet

App Flow - Paid solution with very expensive plans given the low update limits. No API to upload built web directory on normal plans, not very flexible to binaries built elsewhere

### Description

Over the Air Updates / Hot Code Push is the ability to update the js/css/html of your app in real time without having to go through the app store and waiting for reviews. This allows you to push critical fixes out quickly inline with your web app deployments instead of waiting days for Apple to approve entire binary updates.

It would be great if a plugin could facilitate over the air updates from a server/endpoint of our choosing where the developer controls the hosting of the OTA updates. 

The key to the plugin is the ability to download a zip file of your updated dist folder, extract it, tell Capacitor to launch from this new directory until told otherwise. 

There are other factors like verifying updates, choosing when to install them (forced on launch or next time the app is given focus after downloading etc), and optionally downloading only what has updated.

**Comment by mlynch** [comment]
For those looking to do remote app updates in their Capacitor apps, Appflow is a great choice, and it _does_ have an API: https://useappflow.com/

If there's anything we can do to improve the Capacitor + Appflow experience please let us know.

Otherwise, leaving this open for discussion on the topic of microsoft code push.

**Comment by amakh** [comment]
We were willing to give a try to Appflow but even the 25k live updates / month (on the highest plan) is very limited considering that 1 live update = 1 device. In our case, that makes only 2 updates per months (for now, soon only 1) and maybe that could be enough but we don't want to be in the stress that we don't have a 'second chance'.
And we won't be using any other Appflow features to be honest so neither the Entreprise plan is good to us  :/

So any plugin that could just do this job is very welcome 👍

**Comment by lovetingyuan** [comment]
We just need a pure plugin that offers API to manage(update, delete, add) web assets.

**Comment by alexcroox** [comment]
I think updating/deleting assets in use might cause all sorts of issues with a running app so it's probably safer to move the downloaded files into a new directory and set the app to launch from that new folder ([Plugins.WebView.setServerBasePath](https://github.com/ionic-team/capacitor/blob/9788c363c85d372808ff7e21b70421de9ae3a369/ios/Capacitor/Capacitor/Plugins/WebView.swift)?)

A good point though, a light touch native layer with an API to handle just the above, leaving the JS layer and the developer to handle the downloading and version checking etc themselves would great (and quicker/easier to develop).

**Comment by imhoffd** [comment]
@alexcroox Those WebView methods are being pulled into Capacitor core for Capacitor 3, so the usage in native and JS may change slightly, depending on if we try to make it backwards compatible, but they'll still exist. Pretty sure they're all that's needed in Capacitor itself, but someone would need to build plugin(s) that use them.

**Comment by PIXPOINT** [comment]
Maybe for some inspiration for a new Plugin. In the repository [areo/capacitor-codepush](https://github.com/areo/capacitor-codepush) something is used like `setServerBasePath` as @alexcroox suggested.
But unfortunately I didn't get areo/capacitor-codepush up and running.

https://github.com/areo/capacitor-codepush/blob/47a32309ae00f767f46a54ab6a807ebd9721f62c/android/src/main/java/com/microsoft/cordova/CodePush.java#L411

**Comment by alexcroox** [comment]
After much experimentation these last couple of evenings I've made some progress on working this out:

1. Find out the current install path (just so we have a baseline)

`serverBasePath = await Plugins.WebView.getServerBasePath()`

In my case with the simulator this is:

```
{"path":"/Users/alex/Library/Developer/CoreSimulator/Devices/DFB2C8C5-2B02-4071-91DA-E2B5EA8FA02A/data/Containers/Bundle/Application/6279F51C-0757-4BBA-B7D6-E8D611EEC26D/App.app/public"}
```

2. I built another version of the app and zipped up the `ios/App/public` folder. I then uploaded this to S3 to test the download.

3. Download zip from S3 using Plugins.Http

```
const fileDownload = await Http.downloadFile({
        url: `https://xxxxx/app-updates/2.1.1.zip`,
        filePath: '2.1.1.zip'
      })
```

Not providing a directory seems to put it in a `Documents` folder by default. 

```
file:///Users/alex/Library/Developer/CoreSimulator/Devices/DFB2C8C5-2B02-4071-91DA-E2B5EA8FA02A/data/Containers/Data/Application/83EF3EAE-1426-47B5-B4EE-56B37CB66991/Documents/2.1.1.zip
```

4. Using `cordova-zip-plugin` I extract this to the same Documents folder `...Documents/2.1.1/`

5. I now tell Capacitor to permanently launch from this new folder instead of the app/public one:

```
Plugins.WebView.setServerBasePath({ path: '/Users/alex/Library/Developer/CoreSimulator/Devices/DFB2C8C5-2B02-4071-91DA-E2B5EA8FA02A/data/Containers/Data/Application/83EF3EAE-1426-47B5-B4EE-56B37CB66991/Documents/2.1.1' })

Plugins.WebView.persistServerBasePath()
```

And voila the webview reloads automatically and I'm looking at the version of the app I have just downloaded (verified by including some different text in the UI)! 🚀

**But now I'm stuck. Force quitting the app and re-launching no longer loads either app version, and just hangs on the splash screen without error. I can't work out where I went wrong in persisting it.**

**Comment by levarberry** [comment]
If I may...  This is more a business solution than a technical solution.   AppFlow "works".    No it is not free like CodePush.  However, CodePush doesn't work that well with anything other than React Native.  MS has a different business model and motive for giving away CodePush. @mlynch  is not in the same position.

I too was once upset about the price difference.   I expressed it in blogs, forum, twitter, slack .. everywhere.  FREE vs $50/month.  Max changed his per app pricing to allow for smaller groups to be included without having to spend as much.  He made a comprise. $50/app/month was extreme but again he and the Ionic team saw our pain and made the adjustment.

For as much as Max and his team has given to us for $0.  Is $49/month really make or break that we'd rather give it to Amazon/Azure directly?  Wouldn't we rather give it to someone who will turn that into something that will directly give you a competitive advantage? 

What if?  There was a plan below the 10k?  Buy extra Deploys(i think there is already)?  Rollover unused Deploys? 

To cut back on usage .. only use it on iOS since Android approves in <2hrs. 

Lastly there could be a DIY version where you build and host the artifact/manifest yourself and the Ionic Deploy API can read a manifest to determine when/if to download and apply it.      This increases the risk you'll BRICK your app or it could get comprised.

**Comment by alexcroox** [comment]
CodePush works great with our old Vue Cordova app so I'm not sure what you mean by the React Native comment.

I got in contact with an Ionic rep to discuss enterprise costing because our user base (majority free) would quickly fall out of the low update cap per month the $49 offers. We were looking at $8,500 a year minimum for AppFlow enterprise which isn't possible for our startup, which is why we are looking at open source alternatives.

**Comment by levarberry** [comment]
@alexcroox Thanks for bringing up your story.  This is the type of dialogue that would be great for the Ionic team to hear. 

First,  I said "works well" with RN.  It does work with Cordova now.  There was a long gap from the previous version to AppCenter's version support of Cordova.  I still don't find it as smooth as how it works with RN.

Here's more of my opinion on their "low deployment numbers".   I'd side with everyone in saying the number is low.  I'd hope to have all my apps with more than 10K-25K users.   It would be easy to just do a Native App Store deployment ONLY if I changed a native feature and let live Deploy do the rest. 

BUT! is that really a good idea? .. idk.  

What about this solution... it is a bit more work.. but would help keep the the number needed down to a minimum.  
  - Again .. Deploying to Android/Google is only 2hrs so could reduce the number of deployments if it was iOS only.
 - Use Live Deploy as a Stop Gap.    Live Deploy & Native Deploy at the same time.  Limit  the Live Deploy to only lesser versions.

WIth this approach your App Store "LastUpdate" date is real. Your store version numbers are real.  You minimize the need to rely 100% on over-the-air updates, but still able to get out critical fixes ASAP.   Only active users in that 1-3 Day window will need to pull the Live Deploy.

**Comment by mlynch** [comment]
Thanks @levarberry. We're also working on a bunch of new updates to Appflow that should make Capacitor development even better. Also something to keep in mind with Appflow pricing: we're not charging you for the number of updates you make a month (in this case, one or two), we're charging on the delivery of those updates to your users which is what costs money (i.e. bandwidth costs).  I think we'll see Appflow become much more interesting to Capacitor developers than anything coming from Microsoft as they are not focused on web devs building mobile apps.

That being said, will definitely take this feedback into consideration as we think about how to better fit Appflow w/ the needs of the market.

**Comment by SalahAdDin** [comment]
I found this [package](https://www.npmjs.com/package/in-app-updates-capacitor) but its repository was deleted.

**Comment by gustavopch** [comment]
@mlynch Prices that fit in the USA where GDP per capita is about 60k USD don't always fit in countries like Brazil where GDP per capita is about 8k USD. :/

**Comment by ap1969** [comment]
@alexcroox  There was another plugin for cordova at https://github.com/nordnet/cordova-hot-code-push that might be worth a look at for inspiration.

I *think* it installs the new version into a new subdirectory, as you have done, but then copies/moves it into the default directory that the installed apk/ios app uses. (I'm working from memory from looking into this a few years back).

Hope that helps - I'm not an ios/android developer, but would love a solution to this. I've been using Appflow, but the inability to get a good view of what's going on under the hood is causing me problems with an android build that's failing, and Ionic's support didn't spend any time looking into it even though I supplied the build reference, so I'm done with them.

Andy

**Comment by ap1969** [comment]
Further to this, I've tried the line that @alexcroox used earlier (_serverBasePath = await Plugins.WebView.getServerBasePath()_) to get the html directory, and all I get is "public", not the full resolved path.

I've tried: 
Logger.debug(Paths.get(".").toAbsolutePath().normalize().toString());

and I get "/"

I've looked in Android Studio's Device File Explorer, and can't see any of the ionic app files in /data/data/com.app.name/ subfolders.

Anyone got any idea how to view the files?

Regards,
Andy

**Comment by alexcroox** [comment]
@ap1969 it's been a while since I've looked at this but I seem to remember `Plugins.App.getLaunchUrl()` as another option for viewing paths

**Comment by GalvinGao** [comment]
Just tried with AppCenter and, although the project hasn't called for support, it seems that by using their Cordova plugins AppCenter could work. I haven't tested with Code Push yet, but Analytics seems to be working.

See my comment at https://github.com/microsoft/code-push/issues/615#issuecomment-761651376

**Comment by lincolnthree** [comment]
Has anyone seen this plug-in? Looks like someone has most of the basics coded (I have not tested it), but it looks simple enough:

https://github.com/deveshmishra34/penpencil-app-update

We too are in a similar situation. Appflow would be fantastic to use, but we have wayyyy more users than the pricetag allows. Problem is, our app is freemium (doesn't make much money yet) and we can't afford custom/enterprise pricing generally. 

I'd happily pay some regular fee for appflow membership, updates, if I could host and pay for update bandwidth myself without the exploding cost of per-update pricing on our mostly free / unpaid app.




Anyway, that's why I'm here.

**Comment by GalvinGao** [comment]
> Has anyone seen this plug-in? Looks like someone has most of the basics coded (I have not tested it), but it looks simple enough:
> 
> 
> 
> https://github.com/deveshmishra34/penpencil-app-update
> 
> 
> 
> We too are in a similar situation. Appflow would be fantastic to use, but we have wayyyy more users than the pricetag allows. Problem is, our app is freemium (doesn't make much money yet) and we can't afford custom/enterprise pricing generally. 
> 
> 
> 
> I'd happily pay some regular fee for appflow membership, updates, if I could host and pay for update bandwidth myself without the exploding cost of per-update pricing on our mostly free / unpaid app.
> 
> 
> 
> 
> 
> 
> 
> 
> 
> Anyway, that's why I'm here. 

Agreed and same here. We are just a totally-free community website who just want to use the mobile app to deliver more of a better UX. We are also OK to pay for some acceptable membership price but currently the pricing plan is just gonna cost us too far away from what we could afford with...

We are also willing (and even prefer to - given that majority of our user is in China, the CDN out there are probably really slow in the internet environment here without a targeted optimization) to host our own hot update assets, paying the bandwidths ourselves. I know that the Ionic team and Capacitor team may have devoted tons of their spirit in this (and I appreciate that too - AppFlow seems to have a nice looking frontend UI), but again, given the current pricing plan, we are just unable to afford it at all.

**Comment by ap1969** [comment]
I gave up. This issue, and a total lack of interest from ionic when I raised an issue with the build of my android app on their Appflow service (despite me paying for it), made me look elsewhere. I looked at react native (nightmare in this area), nativescript (they have an OTA package in beta, but the creator has gone AWOL for 3 months). And I'm now back on Cordova with Quasar front end and MS code push. It works.

**Comment by lincolnthree** [comment]
@GalvinGao If you try that plugin, I'm interested to hear what your experience is like. As most things open-source, I'm guessing it would take some work, but I think a group effort could make it feasible. We're going to start looking into it in a month or so once we get to the "update our app-store apps" phase of our launch.

**Comment by GalvinGao** [comment]
> @GalvinGao If you try that plugin, I'm interested to hear what your experience is like. As most things open-source, I'm guessing it would take some work, but I think a group effort could make it feasible. We're going to start looking into it in a month or so once we get to the "update our app-store apps" phase of our launch.

I'll take a look on that recent days. That's being said, I also had a success (probably just a lucky one) with the AppCenter plug-ins (but not yet officially supporting Capacitor), mentioned [here](https://github.com/microsoft/code-push/issues/615#issuecomment-761651376).

**Comment by leo6104** [comment]
I made `capacitor-code-push` plugin prototype with `Capacitor 3.0.0-beta.1` + `cordova-plugin-code-push@2.0.0` 

https://github.com/mapiacompany/capacitor-codepush

It forked from https://github.com/areo/capacitor-codepush and merge `cordova-plugin-code-push` master 2.0.0 branch.
And doing some changes to make it available to install on Capacitor plugin.

### Install
```
npm i https://github.com/mapiacompany/capacitor-codepush -D
npx cap sync
```

it also require these plugins
```
    "@capacitor/core": "^3.0.0-beta.1",
    "@capacitor/device": "^0.5.1",
    "@capacitor/dialog": "^0.4.1",
    "@capacitor/filesystem": "^0.3.1",
    "@capacitor-community/http": "^0.2.1",
```


### Configuration

Add your deployment keys to the `capacitor.config.json` file, making sure to include the right key:
```
"Plugins": {
    ... (other plugins)
    "CodePush": {
      "IOS_DEPLOY_KEY": "ANDROID_DEPLOYMENT_KEY",
      "IOS_PUBLIC_KEY": "APP_SECRET_KEY",
      "ANDROID_DEPLOY_KEY": "IOS_DEPLOYMENT_KEY",
      "ANDROID_PUBLIC_KEY": "APP_SECRET_KEY",
      "SERVER_URL": "https://codepush.appcenter.ms/"
    }
}
```


```
import { codePush } from 'capacitor-codepush';
// codePush.sync();
// codePush.checkForUpdate(console.log);
```

### Functions
- [x] Android `codePush.checkForUpdate` 
- [x] Android `codePush.sync()` in first moment. (localPackage.handleCleanDeployment logic)
- [x] Android `codePush.sync()` in second moment. (localPackage.handleDiffDeployment logic) 
- [x] iOS `codePush.checkForUpdate`
- [x] iOS `codePush.sync()` in first moment
- [ ] iOS `codePush.sync()` in second moment

Please try this plugin and Feel free to create PR.
I hope someone can do more progress with my repository. 

### How to deploy 
use command 
```
appcenter codepush release -a Id/AppName -c www/ -t VERSION -d Production -k config/code-push-private.pem
```


### TODO
- [x] Find `CDVWebViewEngineProtocol` / `CDVViewController` alternative in Capacitor and use them in iOS
The failed statement is here.
```
        NSString* startPage = ((CDVViewController *)self.viewController).startPage;
```

**Comment by alexcroox** [comment]
Fantastic effort so far @leo6104 thank you!

**Comment by lincolnthree** [comment]
Hey @leo6104, this looks like a great start! Thanks for sharing.

Out of curiosity, it looks like this requires an AppCenter account - looks like they offer a quite reasonable subscription that includes "Distribute: Unlimited distributions and users"! We'll definitely look into this in a few weeks when we get into this part of the project.

Unfortunately we aren't iOS developers, either :/ It looks like that is the last remaining open issue/blocker on your checklist, is that right?

What is the feature that's not working on iOS? I'm not sure why it's referencing Cordova APIs? `CDVViewController`

Is this code useful from the plugin I linked above?
https://github.com/deveshmishra34/penpencil-app-update/blob/master/ios/Plugin/Plugin.swift
https://github.com/deveshmishra34/penpencil-app-update/tree/master/ios/Plugin

**Comment by leo6104** [comment]
@lincolnthree 

> Out of curiosity, it looks like this requires an AppCenter account - looks like they offer a quite reasonable subscription that includes "Distribute: Unlimited distributions and users"! We'll definitely look into this in a few weeks when we get into this part of the project.

It requires AppCenter account and all free to use code-push deployment in AppCenter.

> Unfortunately we aren't iOS developers, either :/ It looks like that is the last remaining open issue/blocker on your checklist, is that right?

Right! it only have an issue with iOS

> What is the feature that's not working on iOS? I'm not sure why it's referencing Cordova APIs? CDVViewController

I share you current status. 

### `capacitor-codepush` Plugin Status
For Android, it works in common case. I will use this plugin in my production project (two application we have). And continue to manage the project to be stable.
For iOS, Capacitor doesn't inject `CodePush` variable to `import { Plugins } from '@capacitor/core';` so it doesn't work.

### Problems in Android
https://github.com/ionic-team/capacitor-plugins/pull/233 @capacitor/file-system has an issue with copy method. After it merged & released, it will be resolved.

### Problem in iOS
Now, it has Cordova implementation. but It detected to Capacitor plugin. So, Capacitor doesn't read CDVPlugin implementation in the code. It seems Capacitor CLI only detect CDVPlugin implementation within `capacitor-cordova-ios-plugins` folder. they only search CAPPlugin for Capacitor plugin. 

According to Capacitor CLI implementation, if package.json has "capacitor" value, it stop to extract `plugin.xml`. 
https://github.com/ionic-team/capacitor/blob/e148e67c6061be35db361496444d5fa85544a068/cli/src/plugin.ts#L90-L109
It will make Capacitor search Plugin.swift file and CAPPlugin implementation. (Not CDVPlugin implementation) 

### Soltuion
Reimplement all CDVPlugin method to CAPPlugin code pattern. 
```
@interface CodePush : CDVPlugin

- (void)getServerURL:(CDVInvokedUrlCommand*)command;
- (void)getDeploymentKey:(CDVInvokedUrlCommand*)command;
- (void)getNativeBuildTime:(CDVInvokedUrlCommand*)command;
- (void)getAppVersion:(CDVInvokedUrlCommand*)command;
- (void)install:(CDVInvokedUrlCommand *)command;
- (void)preInstall:(CDVInvokedUrlCommand *)command;
- (void)isFailedUpdate:(CDVInvokedUrlCommand *)command;
- (void)isFirstRun:(CDVInvokedUrlCommand *)command;
- (void)isPendingUpdate:(CDVInvokedUrlCommand *)command;
- (void)restartApplication:(CDVInvokedUrlCommand *)command;
- (void)getBinaryHash:(CDVInvokedUrlCommand *)command;
- (void)getPackageHash:(CDVInvokedUrlCommand *)command;
- (void)decodeSignature:(CDVInvokedUrlCommand *)command;
- (void)getPublicKey:(CDVInvokedUrlCommand *)command;
```

**Comment by leo6104** [comment]
Success to implement all iOS CAPPlugin version. 
<img src="https://user-images.githubusercontent.com/7777929/106946634-b1ad8500-676c-11eb-8fe4-c21f4c799e3e.png" height="400">

`codePush.sync()`
`codePush.checkForUpdate()`

**Comment by lincolnthree** [comment]
@leo6104 This is fantastic! We'll try it as soon as we can!  Nice documentation, too!

**Comment by lincolnthree** [comment]
FYI. For anyone finding this thread. The AppCenter CodePush plugin for Capacitor is WORKING and in active development. Please try it out here:

https://github.com/mapiacompany/capacitor-codepush

**Comment by nonsenseless** [comment]
@mlynch Per your request for feedback, the thing I like least about AppFlow is that it forces you to upload your source code to a VM and hope to god that build process understands your directory structure well enough to build it. If it doesn't, you have to futz with it repeatedly to create a build script that will get everything in the correct shape to survive the process.

App Center is a substantially better experience (IMO) because you can manage your build process locally and then you just have to point it at your www folder and let it go to town.

**Comment by mlynch** [comment]
@nonsenseless thanks for the feedback. You can work around this in appflow by doing the build locally and checking in those compiled files.

We will definitely consider making this easier. With App Center's uncertain future we want to make sure teams can migrate to Appflow and not have to significantly change their workflow.

**Comment by nonsenseless** [comment]
@mlynch Is this documented anywhere? You're saying I could check in the www folder and have that processed by AppFlow somehow?

**Comment by mlynch** [comment]
Yes you can, see this: https://ionic.io/docs/appflow/cookbook/appflow-build

Basically you will check in `www` and then set this `appflow:build` script to "true" which means appflow won't do a build but you can keep your regular `build` npm script. 

Then appflow will use the builtfiles in your `www` folder as it would after a normal build.

**Comment by phuthuycoding** [comment]
@leo6104 Hi Leo I have fork repo of you and edit some function to work on capacitor v2. I tested on android , this is work good. But on ios i can't not call setServerBasePath function on `CAPBridgeViewController`, I found issues because `CAPBridgeViewController` not @objc class and `setServerBasePath` is not objc method.

How i can call this method or something else same method on capacitor v2.

Anw, can you create branch 2.x of your repo, I want develop with you. I have create pull request on capacitor, If this pull request to be merge by capacitor issues is resolve

https://github.com/ionic-team/capacitor/pull/4504

**Comment by skjece** [comment]
> Thanks @levarberry. We're also working on a bunch of new updates to Appflow that should make Capacitor development even better. Also something to keep in mind with Appflow pricing: we're not charging you for the number of updates you make a month (in this case, one or two), we're charging on the delivery of those updates to your users which is what costs money (i.e. bandwidth costs). I think we'll see Appflow become much more interesting to Capacitor developers than anything coming from Microsoft as they are not focused on web devs building mobile apps.
> 
> That being said, will definitely take this feedback into consideration as we think about how to better fit Appflow w/ the needs of the market.

@mlynch isn't it possible for Ionic to provide a plugin that allows developers to alter/delte/add files in public/asset folder. We can host the files  (JS,HTML,CSS,Images) ourselves on Github.  That way it can be used free of charge as you won't need to host anything.

**Comment by lincolnthree** [comment]
@skjece Just to play devil's advocate, there's nothing stopping them from doing it, but there's also no incentive for them. They need to make money and they've, reasonably, provided plugins that work with their very painstakingly built tools and services.

While I do think the price of this service from Ionic & AppFlow is extremely / unreasonably high, which presents a barrier to entry that most small-mid developers cannot afford, there is also an open-source alternative you can use *today* if that cost is a deterrent:

https://github.com/mapiacompany/capacitor-codepush

The Capacitor CodePush plugin is actively being developed and we welcome contributions if you find something wrong or would like to add a feature! Microsoft AppCenter also offers the FREE code push to unlimited number of devices, any number of times that you are looking for. It takes setup that Ionic would do for you, but that's the trade-off :)

Cheers and thanks again to the Ionic team for a great open source framework and tools!

**Comment by mlynch** [comment]
One thing to keep in mind: there's a good chance Microsoft will decide to sunset codepush which is another reason we won't be supporting it.

They've been extremely quiet on the app center front, with the only communication being to say they've completely frozen their roadmap to fix deep technical issues. That is a very ominous sign, especially since their last big update was to sunset their push notification service and it's unclear where visual studio cloud services fit into the Azure strategy. Even looking at their Twitter, they've gone almost completely radio silent!

Obviously I'm biased but Appflow will always be the best cloud solution for Capacitor because it's built by most of the same people who make Capacitor. And it has a bright future since we are investing considerably in it. I can't say the same about what I'm seeing with Appcenter.

On the pricing side, we are actively exploring ways to make it better aligned for what indies need. Stay tuned...

**Comment by lincolnthree** [comment]
@mlynch yeah, I've heard that is on the possibility spectrum as well. If that flow pricing were more "Indy"-friendly as you hinted, I would be very interested :)

One more thing to mention about the code push plugin is that you can specify your own server and implement the basic endpoints to do release distribution yourself. Obviously I can't really recommend that unless you want to spend a lot of time testing your service, and then you are also obviously on the hook for bandwidth, but it is an option if one really can't afford a reasonably priced service :-)

**Comment by tolutronics** [comment]
Yes if the pricing is friendly i would prefer Appflow too

**Comment by jmellicker** [comment]
> Yes you can, see this: https://ionic.io/docs/appflow/cookbook/appflow-build
> 
> Basically you will check in `www` and then set this `appflow:build` script to "true" which means appflow won't do a build but you can keep your regular `build` npm script.
> 
> Then appflow will use the builtfiles in your `www` folder as it would after a normal build.

@mlynch it would be wonderful to just build locally, and not have to build in AppFlow, but I found the documentation insufficient for me to implement this workflow.

Could you provide a step-by-step guide? (I'm using Vue/Quasar/Capacitor, and "Ionic Remote" as a repo for OTA updates).

**Comment by nonsenseless** [comment]
@jmellicker You'll need to do the equivalent of 

`
git add www
git commit -m "Added www folder"
`

You'll need to check your .gitignore to make sure that the www folder isn't being ignored (it probably is). At that point, your www folder is checked into git and your AppFlow pipeline will be able to read from it.

**Comment by jmellicker** [comment]
Thanks @nonsenseless !

**Comment by GalvinGao** [comment]
@mlynch Yeah Appflow also looks pretty appealing for me, but the pricing is just simply unaffordable as an open-source project here. Also a majority of our users are located in China Mainland, in which connectivity to foreign servers is an issue, so a possibility of having a self-hosted instance of Appflow would be a huge plus! Looking forward for the new pricing plan :D

**Comment by Silventino** [comment]
@mlynch Appflow prices are waaay over the top. I think a free tier or a cheaper plan with less features would be awesome.

**Comment by riderx** [comment]
@mlynch this subject is coming over and over, please do something for INDIES.
Do a presale put a price who fit and see if you make more money than current one, if not refund us and tell the true, it will not work.
Same as Expo : https://expo.dev/pricing
29 euro by month for appFlow "push update" for solo, I paid blindly, and many here will do too. 

As solo funder I don't need CI, I have Github Action for web and I do native build myself but not often and it's useless my changes are web based 99% of the time .

Have Github Action who create build and use Appflow to tell the app to download it in github it enough 

29* 100 000, is better than 499 * 10
That the final step to make all other Web to native solution useless

**Comment by mlynch** [comment]
Hi all, as I suspected we'd see happen soon, Microsoft just announced [the end of Cordova support for Code Push](https://devblogs.microsoft.com/appcenter/announcing-apache-cordova-retirement/). Given they paused all feature development on App Center last year and have had a stream of service shutdowns and retirements, we don't think it makes sense to invest in further integration with App Center. I suspect what will happen with App Center eventually is it will be sunset to get developers to move to corresponding Azure services (if there are any), and these shutdowns are just the start.

Of course, the community can fill in this gap but the future for App Center doesn't look very encouraging. We are trying to evaluate the impact of this shut down to see if we can make it easier for developers using Code Push to move to [Ionic Appflow](https://ionic.io/appflow) which has superior remote app update features for Corddova and Capacitor, along with native builds and app store publishing. Stay tuned for more info on that when we have it.

**Comment by o-alexandrov** [comment]
@mlynch The mentioned AppCenter news has nothing to do with using Capacitor.
- AppCenter Cordova wasn't meant for Capacitor and was a feature in preview

While there is no active development to the previously created solution, you could see issues are promptly addressed in AppCenter's repo as well as they mentioned ongoing work.

Some references how you can configure AppCenter w/ Capacitor (as a native iOS / Android project):
- https://arielhernandezmusa.medium.com/building-ionic-capacitor-on-appcenter-fe0105bb02d1
- https://www.kevinboosten.dev/release-your-ionic-and-capacitor-app-via-appcenter

Using Appflow is financially unreasonable for any company, unless it's a small scale internal tool.

**Comment by mlynch** [comment]
@o-alexandrov this news from Microsoft is absolutely significant and is now a pattern of service sunsets and shutdowns impacting App Center. As an outsider, it's clear that Azure is the future for cloud at Microsoft. As a service under the Visual Studio brand, it's not clear where this fits in Microsoft's strategy. Again, feel free to integrate with it but at your own risk. We won't be taking that risk ourselves.

As for Appflow, happy to chat more about how we can make it work for your team. We just rolled out new [Community and Startup pricing](https://go.ionic.io/community-plan-faq) for indies and startups!

**Comment by o-alexandrov** [comment]
@mlynch thank you for the added pricing options.
However, personally, I wouldn't suggest any of my clients to use it even based on the updated pricing.

With an exception, it's great to support Ionic hoping it'd further improve Capacitor.js.
Please consider adding the following:
1. The distribution needs a pay-per-use model based on the computing minutes spent
    - at least similar to GitHub Actions
2. Over-the-air (OTA) updates needs a self-hosted option
    - letting devs to upload assets on, for example AWS S3, and pointing AppFlow to the uploaded version

**Comment by mlynch** [comment]
Sorry to hear that. We think it's a awesome service and we have thousands of customers on it and hundreds of enterprises that really, really like Appflow. Since it's built _for_ Capacitor (and Cordova) and by many of the same people who work on Capacitor, the level of deep integration we're able to have is second to none and we're going to be adding more and more features that will take advantage of this deep integration soon.

If you ever want to give it a more thorough try let me know and I'd be happy to set you up with a longer trial!

Also we do support self-hosted assets for live updates and have a number of teams using that feature.

**Comment by riderx** [comment]
@mlynch Your Community and Startup pricing for indies and startups don't include Over-the-air update, your plan do not make sense for indie, tiny team do already build the app themself and will continue, what we cannot do is the OTA, who is missing.

OTA with self-hosted assets is the only feature i need from appflow

**Comment by mlynch** [comment]
@riderx apologies, the Startup pricing _does_ include live updates, currently the community one only includes native builds and not live updates.

**Comment by riderx** [comment]
@mlynch so indie are still stuck, 49 is very expensive for indie and do not include what we need...
Please take  time to think about it.

**Comment by amakh** [comment]
If we use self-hosted option, do we still need to subscribe to a plan and pay when we exceed the limit?

**Comment by Silventino** [comment]
> @mlynch so indie are still stuck, 49 is very expensive for indie and do not include what we need... Please take time to think about it.

Actually, it is $249 per month lol
And you have to pay a whole year upfront........

**Comment by levarberry** [comment]
I'll say this  because I know Max can not.

You have to ask yourself why as an Ionic developer you can't afford $49.
What is it about your business or career that doesn't allow you to spend
$49 on a tool that can help you?     The amount of time and effort saved
is tremendous.      You probably spend $40 on hosting and spend $20 on
canva and a lot of misc money on other things.  $49 or even $200 is not a
lot of money to ask.     If you don't have a Mac and need Native builds ...
pay someone to build them for you.   If you can't really can't afford $49
for live updates .. that means to me your users are not paying you any
money and therefore should wait on App Store updates.

In addition...  if  Ionic was to charge anything less ...  I would be in
fear that this project would fall because of lack of funding.    Ionic is a
business not a charity.    I run a business not a charity, so paying helps
me feel like they are going to put the effort in to keeping this alive.

Open your eyes to what has happened with all the other FREE / Low cost
solutions.    They are gone!

Is it fair to Max to fund a project that costs him a lot of money and
nothing in return just for the sake of people who pay him nothing?    Do
you write Ionic apps for free for your clients?   He's already not charging
for the web components and Capacitor and Stencil... what more do you want
from him?

It doesn't matter what country you are from.  This is not a lot of money.
  If you need money ...   you can always work in my company and I'll pay
your $49.


*Levar BERRY - CEO*
(o) (513)318-1360 ext.101
(c) 513-403-1210
(e) ***@***.*** ***@***.***>
(w) http://edrivenapps.com/

**Comment by gustavopch** [comment]
@levarberry You have a point, it's important that Ionic can fund itself and be profitable. That said, you're terribly wrong here:

> It doesn't matter what country you are from. This is not a lot of money.

The impact of US$249/mo is **massively different** for a company based in a country where GDP per capita is US$60k/year Vs. where it's US$6k/year.

Look at Spotify. They seem to charge US$9.99/mo in the US. Do you know how much they charge in Brazil? US$3.60/mo in the current dollar rate. Is that because people here are penny-pinchers? Of course not. It's because people in this country are not as rich as people in the US. Spotify knows that and they're not a charity. While 50% of people in the US live with US$3k/month or more, it's about 2% of people in Brazil. And Brazil isn't even one of the poorest countries in the world — there are a lot of people in a much worse condition just because of where they were born. Sometimes it really seems like people in 1st world countries can't even realize how well they are compared to the rest of the world.

Please read this: https://ourworldindata.org/poverty-minimum-growth-needed.

Edit: sorry for the rant, but I think it's important for the discussion.

**Comment by mlynch** [comment]
@levarberry thanks for the support, we really appreciate it.

It's hard balancing making money and giving things away for free. I strongly believe that Ionic generating revenue is the best thing for the community. Without it, the services companies rely on are on shaky ground, as we see with App Center. It's great that App Center is free or very cheap, but it's not so great when that means you can't rely on it as a business.

The revenue we generate goes back into our open source offerings to make them better. We know not everyone will choose or be able to be a paying customer of our commercial services but I do think it's important to point out where that money ultimately goes. 

Also we very much understand prices can be felt very differently in different countries. However, we also operate in an environment where we are paying US prices so there's a limit to how low we can go before something simply becomes not economically viable for us. I am guessing Microsoft had been losing money on Code Push bandwidth for a long time but kept it around as a loss leader. Seems that bill is coming due. We won't have that issue as we've priced our product with our costs fully in mind to make sure it's sustainable.

**Comment by lincolnthree** [comment]
Since I've been active on this thread, I'll just leave two main comments/ideas here.

Before I get into that, though, I will say that the folks at Ionic have worked very hard on providing solid tools for developers to use, and they have have been very generous with open-sourcing most of them and *supporting* the open-source community. Thank you. You guys should definitely charge money for this service, and your new pricing is MUCH more fair than what I think the prior model presented as a barrier to small/indie devs.

Okay...

First, as a proponent of open-source, and someone who's worked on the Capacitor plugin for AppCenter/CodePush, I will say this. If you are not able to pay for a $249/mo (~$3000/yr) subscription for this service, then please consider supporting this project, and consider starting an open-source project implementing the AppCenter server/API endpoints:

https://github.com/mapiacompany/capacitor-codepush

AppCenter and the APIs required to make a functional code push server are NOT complicated. Building a UI to make it all pretty and easy to use would be a bit more work, but if you are willing to implement the AppCenter APIs, and upload a few files for clients to check / download when you want to push out an update, then this would be very straightforward to implement and host yourself.

The `capacitor-codepush` plugin makes only two or three API calls, roughly, including the download of the new distribution, and this could fairly easily be reproduced in an open-source solution. 

1. Client calls to check for availability / metadata of new update - provides its current version and device info.
2. If update is available, client requests download of zip file including new code.
3. Client calls to notify server update was successful or if rolled back. (This could be ignored if you want to run wild and don't care about tracking errors.)

Second, I'd also suggest contacting Microsoft support, as per their suggestion, and asking them to open-source their Cordova codepush solution.

> "If you have an Apache Cordova app and need assistance evaluating other options, our team is more than happy to assist you via our support channel. Simply open a support ticket under the drop-down menu on the right top corner of the portal."

**Comment by gustavopch** [comment]
@mlynch To clarify my point, I don't think simply lowering prices is necessarily a sustainable solution. I can absolutely see how US$249/mo for a company based in the US can be quite cheap if you consider how much time would be spent building all of that from scratch. Country-based pricing on the other hand may make it affordable in places where it isn't today. End result may be growing the revenue by reaching potential users that are not reached today.

**Comment by riderx** [comment]
@levarberry thanks for your comment, you said many true things, but you forgot many point.
I do build app that are free to use and open souce, in your exemple i don't have the right to have great tool, a bit unfair.
Second point, 49 dollar is expensible for indie but doable as i said. But in this price we don't have the only interesting feature, the code-push .
I don't know any indie who need CI/CD for building since you are alone you build on your machine, want you cannot do is sending OTA update.
@mlynch thanks you made some effort and to keep the dialog open, but this offer do not fit the need, in my point of view.
Even when i was in startup less than 1M i don't see the point.
In the other way around for sure 49 for only OTA and no CI/CD look smarter, of course when the team got bigger i will need your CI/CD 100% of the time .

**Comment by riderx** [comment]
I'm on my way to build a module who allow you to download update from a given url.
Something like :
```
import { updateApp } from 'capacitor-updater'
updateApp('URL_TO_S3_OR ANY_PLACE')
```

It will try to :
- download data from url
- unzip data it
- replace the current code with the new one

The plugin will be simple, you manage on your side the update timeline and when to update.
If you are interested put emoji or send me email to discuss about it.

**Comment by ap1969** [comment]
I'm also dusting of the old nordnet components. CLI is done, running proper tests tonight. I can probably do the java, but it's been a long time since I worked with it, so a second pair of eyes would be great. I've never worked on native IOS apps, so a hand with those would definitely be needed, if anyone is interested?

https://cordovahotcodepush.com for updates.

**Comment by riderx** [comment]
@ap1969 let's join our efforts : https://github.com/riderx/capacitor-updater
I just made the base of Android and IOS needs.
It's still WIP but any help is welcome !
The blocking poitn now is to manage the unziping for downloaded files.
I will continue next week

**Comment by ap1969** [comment]
Does it only work with capacitor? 

To be blunt, I'm avoiding using anything to do with ionic because I hit problems with their Appflow, and got zero help to diagnose the problem. I couldn't even see logs, and they wouldn't send any.

**Comment by lincolnthree** [comment]
If you're looking for a client side module that does the update for Capacitor... it's already built. Just need to build out the server end! (E.g. Just need to swap out the AppCenter URL for your own server -- but we need help with the server / CLI part!)

Here's the client code: https://github.com/mapiacompany/capacitor-codepush

**Comment by o-alexandrov** [comment]
@lincolnthree you don't need a server (one that you would pay for computing expense) to have OTA updates.
Here's a simple scenario:
1. clone an example capacitor.js web app
2. make the example capacitor.js web app to look at a specific URL (for example AWS S3 endpoint) for updates with `version.json` file
3. build an example web app and upload to AWS S3
    - together with the build files, add a `version.json` file

In the proposal above, all you need to determine whether to download a new version is the static `version.json` file:
- `GET https://aws-s3-endpoint/version.json`
  - if a version is different than what you have client-side (on the device), download the build and replace a directory with web app files you have client-side
  - if needed, add authorization layer based on your preference
  - later on, add hooks api client-side, so developers could add their custom logic at least for `onSuccess` (on successful download) and `onError`
    - if needed, on the backend (AWS S3) in this case, a developer can add a Lambda function for AWS S3 events
  
That's what I believe @riderx and @ap1969 are implementing with [capacitor-updater](https://github.com/riderx/capacitor-updater)

**Comment by ap1969** [comment]
Exactly that. I've got android updates working already. Moving onto iOS now.

**Comment by tolutronics** [comment]
> Does it only work with capacitor?
> 
> To be blunt, I'm avoiding using anything to do with ionic because I hit problems with their Appflow, and got zero help to diagnose the problem. I couldn't even see logs, and they wouldn't send any.

@mlynch

**Comment by riderx** [comment]
@ap1969 yes the plan was to make it work in capacitor only, no server need no overload.

**Comment by ap1969** [comment]
Hi @riderx 
Sorry, I meant would your solution work in plain Cordova as well?
Andy

**Comment by riderx** [comment]
@ap1969 i used the capacitor start plugin from here: https://capacitorjs.com/docs/plugins/creating-plugins
So it seems it's not.

**Comment by jfbloom22** [comment]
@mlynch I am a huge fan of Ionic AppFlow.  It has worked flawlessly for me.  The reason I hope this thread continues to work on a community solution is because I am running into the 100 install limit on the Community plan and it does not make sense to spend $500 a month to deploy an app to 30 users. 

I expect there are a lot of developers out there supporting apps for an internal mobile team of 20 - 200 users who will also find $500 too pricy.  

Additionally, the app I am working on is an Enterprise iOS app so we don't publish to the store. I have 0 interest in building the binary on AppFlow.  

It would be brilliant if the Community plan included 500 live updates.  Or better yet, if there was a Live Update only package for $50/month that included 5,000 Live Updates.  Bottom line is that I am happy to pay a bit more than $50 a month for more than 100 live updates, but no way can I justify $500 to distribute an app to 30 people.

**Comment by ap1969** [comment]
@riderx OK. I don't know much about capacitor, hence my asking. That's a shame.

**Comment by ThodorisElissaios** [comment]
@mlynch  i was paying $49/mo just for the live updates, now you are asking me $499/mo ?

Who is the genius behind this move?

Ionic used to be an awesome framework with reasonable priced tools. (i have been using the platform since DragNDrop builder era & Push notifications) 

And  now you are asking $4,999/yr ?

**Comment by nonsenseless** [comment]
@ThodorisElissaios I'm not crazy about their pricing model either but there's no need to be rude. They're trying their best to figure out how to make an actual functional business model and that ain't easy when everybody expects you to give your work away.

**Comment by ThodorisElissaios** [comment]
Please everyone excuse me if i am rude. 
I completely understand they're trying their best to figure out how to make an actual functional business model.

In my opinion asking for an (almost) 1000% increase sounds rude.

**Comment by mlynch** [comment]
Appreciate all the feedback, truly!

@ap1969 sorry you had a bad experience with Appflow, we've made a bunch of improvements to the tool and our support process recently so I hope that you give it another try some day!

@jfbloom22 I can understand if it's not a fit in that case. Most of our customers are using our other features in tandem or in larger companies that need other enterprise features. If you just need a few live-updates and nothing else I agree that's not going to fit with our current plans but we will certainly keep this in mind.

And @nonsenseless @ThodorisElissaios thanks. I understand that charging for things in the open source world can be contentious, but I do want to point out that because Ionic generates revenue, we can actively employ full time engineers to work on projects you love like Capacitor and Ionic. Our business model seems to work best when we don't charge Indies or Students much or anything, but focus on charging serious businesses because they need a lot more and they also need a lot more *support* and security guarantees for everything we sell. That's why I'm legitimately happy if the indie market has a path forward here but I want to be transparent about how we decide to put an official stamp on something.

At the end of the day Ionic is still here and doing well while App Center, which doesn't charge enough, is starting to crumble. It's really hard if you don't have real revenue to point to when it comes time to validate your existence in a big company it turns out.

**Comment by jfbloom22** [comment]
If anyone is in a similar boat and needs a few more live updates on the Community plan, I recommend contacting Ionic support and simply asking.  I asked for 500 live deploys, explained the situation [(above)](https://github.com/capacitor-community/proposals/issues/43#issuecomment-951359381) and they graciously granted my request. 

I have built over 12 solutions with Ionic Framework, starting when it was in beta years ago.  Recently I started to have my doubts because of this poor pricing with Live Updates, but now my confidence in Ionic is completely restored.  Really impressed with the folks over there at Ionic.

**Comment by nonsenseless** [comment]
@mlynch I will say that it would be great to have more transparent pricing on the enterprise-level options. We have quite a few clients and, with all due respect to sales people, there's nothing more frustrating than having to talk to them to get basic pricing information and then navigate an endless series of "JUST WANTED TO FOLLOW UP" emails until one or the other of you finally dies and can stop using email.

**Comment by mlynch** [comment]
@nonsenseless agreed, we rolled out much more pricing transparency recently which has been well received. Open to feedback if we're missing the mark still! https://ionic.io/pricing

**Comment by nonsenseless** [comment]
@mlynch Nice, this is definitely much improved! For the indie/startup plans, what kind of live update limits are there? And what kind of documentation do you ask for from start-ups? We work with a lot of folk developing MVPs so it would be helpful to have a clearer overview of the process and options to give them.

**Comment by riderx** [comment]
@mlynch do you know the pricing of revenue cat ?
https://www.revenuecat.com/pricing
Maybe you can ask user to connect they webhook of apple store and google to check the app revenue and do a limit like revenue cat do ?

**Comment by jamesyoung1337** [comment]
> If you're looking for a client side module that does the update for Capacitor... it's already built. Just need to build out the server end! (E.g. Just need to swap out the AppCenter URL for your own server -- but we need help with the server / CLI part!)
> 
> Here's the client code: https://github.com/mapiacompany/capacitor-codepush

@lincolnthree What exactly is needed here?  Is there some basic documentation for the endpoints and how the cli works?

I would be interested in hosting something for open source and other projects.

Can we verify this plugin works as expected, more or less, and that only the URL needs changing along with the new endpoints and cli?  Since you say it's not overly complex to implement the endpoints save for a pretty interface I would like to get this up and running and perhaps work on a nicer interface later, when it works.

**Comment by lincolnthree** [comment]
@jamesyoung1337 Awesome! Yes, I both helped develop this iteration of the plugin, and use it in production on over 100,000 devices. (iOS and Android).

I think the best documentation is the source code of the client capacitor plugin, which is... typical for OSS, but of course not ideal :) Microsoft liked to keep their server API proprietary, and open-source the client stuff.

Essentially the current Capacitor CodePush plugin makes two API calls:

1. Call: What version is available for me? (`GET /checkUpdate?clientVersion=1.0.0&stage=PROD`), which returns info on available update for a given update stage/track.
2. If an update is available, then it downloads the file (`GET /path/to/1.0.1.zip`) and unpacks it to the device filesystem. If the update succeeds, it can reload immediately, wait until background timeout, or wait until next launch. If it fails, the update is rolled back.
3. Call: Then it informs the server if the update succeeded or failed (`POST /updateResult?version=1.0.2&status=[success/fail]`) - this is not essential and could be skipped. It is used for stats/metrics - and is useful to see if you need to pause/roll back an update.

It's really not much more complicated than any of the new solutions being proposed on this thread. It also already supports verifying the MD5 checksum and signed contents of the archive to prevent malicious updates / hacking. 

So as you can see, there is a lot of great / useful functionality in here that really would be quite easy to re-use with a few tweaks. It would be a shame to lose all of this work.

What is needed is:

- An update/fork of the AppCenter CLI to replace their bundle uploading to build a local bundle, sign, and upload to S3 or GitHub or wherever the file is hosted, and add the new build to the update server/service.
- A service to receive the API calls from the client and do some version comparisons with the available updates and the client app version / return appropriate update info/file URL.
- Updates to the client plugin depending on / made to match the changes above.
- A service to track success/failure of updates (optional)

Anything beyond that, like a GUI to pause/roll back an update is just candy. I think this should be pretty easy to host straight out of GitHub or S3 using a JSON file, and possibly moving the version comparison logic to the client, to be honest.

**Comment by lincolnthree** [comment]
Here are some of the HTTP calls, which I believe happen solely in the TypeScript/JS, and not in native code:

Check for update (code-push package included via NPM):
https://github.com/microsoft/code-push/blob/master/src/script/acquisition-sdk.ts#L97

Download the file:
https://github.com/mapiacompany/capacitor-codepush/blob/main/src/remotePackage.ts#L56

**Comment by jamesyoung1337** [comment]
@lincolnthree Thank you, that is very helpful!

I have been working my way through the source code, and I have the general hang of it as far as the appcenter cli goes.

I will post a link to a github repo here when I have some legwork done.

I'm interested in some details on how you have used in production: that was this plugin and maybe your own backend, localhost or your own server?  I'm working my way through the commit history, will let you know how I get on.

Thanks and talk soon!

**Comment by lincolnthree** [comment]
@jamesyoung1337 My pleasure! Perhaps we should move this discussion somewhere else. Possibly an issue in the Codepush Plugin repo? Happy to share more details off this main thread or via email / Discord: LincolnThree num 1078

**Comment by jamesyoung1337** [comment]
@lincolnthree Yes, let's do that. An issue at [your repo here](https://github.com/lincolnthree/capacitor-codepush)? Discord channel: [Capacitor and CodePush](https://discord.gg/CEcY2ShD)

**Comment by riderx** [comment]
i just update [`capacitor-updater`](https://github.com/riderx/capacitor-updater) it's working on IOS, i improved the DX a bit :
```
  import { CapacitorUpdater } from 'capacitor-updater'

  const testUpdate = async () => {
    const version = await CapacitorUpdater.download({
      url: 'https://github.com/Forgr-ee/Mimesis/releases/download/0.0.1/dist.zip',
    })
    await CapacitorUpdater.set(version)
  }
```

This will download the version and load it.
Edit: hard reload app work too ! 
**Have fun !**
if you need something please open issue

**Comment by riderx** [comment]
I made it !

**Comment by o-alexandrov** [comment]
@riderx pls consider to create some UI around it to offer a commercial version, so you could support your solution long term :)

**Comment by lincolnthree** [comment]
@riderx Do you have a discord server we could all chat on? Possibly set up a server to collaborate? I am happy to get this started if it would be helpful.

**Comment by riderx** [comment]
@lincolnthree you can join here : 
https://discord.gg/VnYRvBfgA6

I'm working on the Android Version

@o-alexandrov I'm not sure what I can do for commercial version.

I will setup sponsorship account, I think it can be an easy solution too.
Feel free to join the Discord to discuss about it

**Comment by riderx** [comment]
Just released [`capacitor-updater`](https://github.com/riderx/capacitor-updater) V`1.0.05` with Android hot reload, like IOS I need to make it persistent between app kill.

**Comment by riderx** [comment]
And Boom Found the way to persistency in Android with `1.0.9` thanks @jamesyoung1337 for the help !

`npm i capacitor-updater` and enjoy, I improved the doc too : [`capacitor-updater`](https://github.com/riderx/capacitor-updater)

**Comment by skydusky** [comment]
https://github.com/riderx/capacitor-updater/projects


 this project works flawless 
its dead simple you got control
dont have to rely on anyone for your code push you could use static S3 or put a api in front for basic analytics
perfect 🔥

**Comment by riderx** [comment]
Thanks @akashdusky the project have still few issue, but the idea was to make it very simple !
For example I didn't find a way to revert the app and use the original code again.
Or handle launch issue, yet. But I will continue to work on it !

**Comment by riderx** [comment]
I foud the way to revert to original app !

**Comment by hoveit** [comment]
Great work! @riderx

**Comment by riderx** [comment]
After the manual mode of [capacitor-updater](https://github.com/riderx/capacitor-updater) to download updates.
I released the auto-update mode, you can connect it to your backend or mine https://capgo.app .
It's all open source and all dead simple, enjoy !

**Comment by skjece** [comment]
> After the manual mode of [capacitor-updater](https://github.com/riderx/capacitor-updater) to download updates. I released the auto-update mode, you can connect it to your backend or mine https://capgo.app . It's all open source and all dead simple, enjoy !

This feature was much needed as ionic charges too much for this . Thanks a lot @riderxx

**Comment by skjece** [comment]
> After the manual mode of [capacitor-updater](https://github.com/riderx/capacitor-updater) to download updates. I released the auto-update mode, you can connect it to your backend or mine https://capgo.app . It's all open source and all dead simple, enjoy !

This feature was much needed as ionic charges too much for this . Thanks a lot @riderxx

**Comment by riderx** [comment]
@skjece ionic charge a lot, i agree, but for way more than live updates.
Native build are a pain in the ass too, and my tool doesn't fix that, when you update native code you have to send a new version in the store.

**Comment by alexcroox** [comment]
If you want to automate that part try fastlane, I've used it for about 5 years now on various projects and it takes all the pain out of auto building and auto submitting to Apple/Android via CI/CD. Past couple of years I've had it running great in GitHub Actions https://docs.fastlane.tools/

**Comment by riderx** [comment]
@alexcroox it like all CI, it's scary to configure at beginning!
For all others who have followed the adventure, https://capgo.app is starting to be paid for usage in 15 days, with a fair price.
You can still use the plugin for free in manual mode or with your own server, if you do don't forget to support open-source : https://github.com/sponsors/riderx <3

Thanks a lot all for feedback and energy it helped me to start this wonderful project!

**Comment by robingenz** [comment]
Hey everyone, 
since this issue is still open I wanted to mention the brand new [Capacitor Live Update](https://capawesome.io/plugins/live-update/) plugin from Capawesome.

**Features**: 

- 🔋 Supports Android and iOS
- ⚡️ Capacitor 6 support
- 📦 Bundle Management: Download, set, and delete bundles.
- ☁️ Cloud Support: Use the [Capawesome Cloud](https://capawesome.io/cloud) to manage your app updates.
- 📺 Channel Support: Set a channel for the app to manage different versions.
- 🔄 Auto Update: Automatically download and set the latest bundle for the app.
- 🛟 Rollback: Reset the app to the default bundle if an incompatible bundle has been set.
- 🚀 Rollout: Gradually roll out new bundles to gather valuable feedback.
- 🔒 Security: Verify the authenticity and integrity of the bundle using a public key.
- 🌐 Open Source: Licensed under the MIT License.

I would really appreciate your feedback. Feel free to create a discussion in https://github.com/capawesome-team/capacitor-plugins if you have any questions or send me an [email](mailto:robin@capawesome.io) (so we don't spam this issue). Have a nice day!

**Comment by SalahAdDin** [comment]
That repo is amazing @robingenz !
